### PR TITLE
Add client user connection limit

### PR DESCRIPTION
**Pull request by AndrewJackson2020**
This PR adds a user level client connection limit. I believe that currently user connections can only be limited by how many people are given connections  to the database backends but there is no way to limit the number of connections to pgbouncer itself. Combined with the global connection limit this means that one user could use up all connections on the pgbouncer instance. This PR proposes a fix to this problem.

Related to #1136 (because it partially fixes it)

### CODE REVIEW

File: include/bouncer.h
Comment (JelteF) [inline include/bouncer.h:543]:
```suggestion
	int max_user_client_connections;	/* how many server connections are allowed */
```

File: include/bouncer.h
Comment (JelteF) [inline include/bouncer.h:545]:
```suggestion
	int client_connection_count;	/* how many client connections are used by user now */
```

File: include/bouncer.h
Comment (JelteF) [inline include/bouncer.h:544]:
```suggestion
	int connection_count;	/* how many server connections are used by user now */
```

File: src/client.c
Comment (JelteF) [inline src/client.c:508]: Can you undo this?

**Review by JelteF — CHANGES_REQUESTED** [review]
This seems like a good solution to the described problem. A similar setting would probably be useful per database, but that doesn't need to be in this PR. 

Some things that should be changed about the PR:
1. Github Actions should be green (fix C warning and fix formatting)
2. A test should be added (`test/test_limits.py` seems like a good place to add one)
3. The new option should be documented in `doc/config.md`

File: src/client.c
Comment (AndrewJackson2020) [reply src/client.c]: Done

File: include/bouncer.h
Comment (AndrewJackson2020) [reply include/bouncer.h]: Done

File: include/bouncer.h
Comment (AndrewJackson2020) [reply include/bouncer.h]: Done

File: include/bouncer.h
Comment (AndrewJackson2020) [reply include/bouncer.h]: Done

**Comment by eulerto** [comment]
I didn't review this PR yet but I noticed that you did lots of changes in the documentation that is not related to this feature. If you want to propose a whitespace cleanup, create a separate PR. We generally use `/* comment */` instead of `// comment`.

BTW, I'm fine with this feature. It is equivalent to `CREATE ROLE ... CONNECTION LIMIT`.

**Comment by AndrewJackson2020** [comment]
> I didn't review this PR yet but I noticed that you did lots of changes in the documentation that is not related to this feature. If you want to propose a whitespace cleanup, create a separate PR. We generally use `/* comment */` instead of `// comment`.
> 
> BTW, I'm fine with this feature. It is equivalent to `CREATE ROLE ... CONNECTION LIMIT`.

Apologies about that, my editor eliminates trailing whitespace default. Will revert momentarily.

**Comment by AndrewJackson2020** [comment]
> I didn't review this PR yet but I noticed that you did lots of changes in the documentation that is not related to this feature. If you want to propose a whitespace cleanup, create a separate PR. We generally use `/* comment */` instead of `// comment`.
> 
> BTW, I'm fine with this feature. It is equivalent to `CREATE ROLE ... CONNECTION LIMIT`.

I reverted the whitespace removal and changed the comment style to conform with the rest of the codebase.

**Review by JelteF — CHANGES_REQUESTED** [review]
Thanks! That looks a lot better already. I left a review just now on your other #1138 PR.  Most of the feedback there should also be applied here. Other than that it looks pretty good.

I still have want to do some manual tests of both PRs though

**Comment by AndrewJackson2020** [comment]
@JelteF Made all of the documentation changes and added a test of the stats command in the unit tests. Also split the unit test out into 2 separate ones: one that is meant to deny a connection and one that is meant to accept a connection, I feel this is a bit cleaner.

A few additional thoughts:
1. Should this PR and #1138 be merged because they are so similar and have similar feedback?
2. Should the name of the variables be changed from max_user_client_connections/max_user_connections to max_user_client_connections/max_user_server_connections?
3. Should the tests of the stats command  be moved to test/test_admin.py or is it fine as is?

File: src/objects.c
Comment (JelteF) [inline src/objects.c:1398]: This substraction should only be done when this specific clien actually increased the connection count before. 

Right now if a user is at the limit. The next failing connection attempt (due to hitting the limit) will then also decrease the connection count. So the connection attempt after that will succeed again, eventhough we're still at the client connection limit.

**Comment by JelteF** [comment]
>     1. Should this PR and [Add db client connection limit #1138](https://github.com/pgbouncer/pgbouncer/pull/1138) be merged because they are so similar and have similar feedback?


>     2. Should the name of the variables be changed from max_user_client_connections/max_user_connections to max_user_client_connections/max_user_server_connections?

Let's keep them as is. I don't think the better naming (because I agree it is a bit confusing now) is worth breaking backwards compatibility of the config setting names.

>     3. Should the tests of the stats command  be moved to test/test_admin.py or is it fine as is?

It's fine as is imho.

File: src/objects.c
Comment (AndrewJackson2020) [reply src/objects.c:1398]: Okay so I added a unit test assert involving the situation you described above. Moved the increment to right before the initial check and changed the inequality accordingly. Seems to satisfy all test cases though there might be another corner case I'm not seeing.

An alternate way of going about this is adding a boolean to each client that specifies if they contribute to the user level count and check that before decrementing the count. I feel like the logic in this commit is better since it adds less variables, as long as the logic is correct.

Another thing that I notice is that this new logic does not apply to PAM authed users though I feel like it should.

Comment (JelteF) [reply src/objects.c:1406]: > An alternate way of going about this is adding a boolean to each client that specifies if they contribute to the user level count and check that before decrementing the count. I feel like the logic in this commit is better since it adds less variables, as long as the logic is correct.

Agreed.

> Another thing that I notice is that this new logic does not apply to PAM authed users though I feel like it should.

Agreed. Everywhere we set `client->login_user_credentials` we should do this incrementing/decrementing correctly. Also the block below where auth_user is used should do this. I think the easiest way to do this is do

```
client->login_user_credentials = /* something */
if (!check_user_connection_count())
    return;

bool check_user_connection_count() {
    /* your check */
}
```

File: src/client.c
Comment (JelteF) [inline src/client.c:484]: Let's add a comment here for clarity.
```suggestion
			/* increment count now, so that we can decrement it safely in disconnect_client if limit was reached */
			client->login_user_credentials->global_user->client_connection_count++;
```

File: src/objects.c
Comment (AndrewJackson2020) [reply src/objects.c:1406]: Encapsulated the logic into the function and added the call to everywhere client->login_user_credentials is assigned. Tests all passed but the surface area of the change is now much bigger which may mean more testing is required (limits correctly incremented, decremented and enforced for pam users, admin users, etc). Also the check_user_connection_count function could probably be refactored a bit (lot of nesting).

Comment (AndrewJackson2020) [reply src/objects.c:1406]: @JelteF Expanded test coverage to include admin users. Refactored the check_user_connection_count function to prefer early returrns instead of nested if statements. Refactored the tests so they should be faster, rely less on asynchronously running pg_sleep.

Could also expand test coverage to stats users as I believe are instantiated differently than admin users. Also missing is test coverage around PAM users though I dont believe there is PAM coverage elsewhere in the test codebase either.

Let me know if anything else needs to be done with this PR in order to get it in a mergable state.

Comment (JelteF) [reply src/objects.c:1406]: If tests pass, I think this is probably good. I'm unavailable for the next ~3 weeks though.

Comment (AndrewJackson2020) [reply src/objects.c:1406]: Understood, thanks for the heads up.

**Review by JelteF — CHANGES_REQUESTED** [review]
Sorry for the delay in reviewing. I think this is almost good now, some final suggested changes.

File: src/admin.c
Comment (AndrewJackson2020) [reply src/admin.c:1683]: Good point. Looking at the implementation of the server wide max_client_conn setting it behaves in the way you are describing. Will update that.

File: doc/config.md
Comment (AndrewJackson2020) [reply doc/config.md:1435]: Sorry about that, will fix.

File: src/client.c
Comment (AndrewJackson2020) [reply src/client.c:550]: Will do. Will also make sure test coverage for this feature includes users from the auth DB.

File: src/admin.c
Comment (AndrewJackson2020) [reply src/admin.c]: This is fixed now. Added unit tests to verify that the connection limit applies to the stats users accessing the admin console but not admin users accessing the admin console.

**Comment by AndrewJackson2020** [comment]
@JelteF Pushed a new commit that fixes a bunch of the issues the issues that you mentioned. 

Found that the auth pass through issue that you mentioned in #1138 applies to this PR as well. Implemented a somewhat hacky fix to this. Would appreciate your feedback on this or if you have any alternative ideas.

Also expanded unit test coverage for stats users and auth pass through users. One area that isnt covered is PAM but looking through the unit tests I dont think there is any test coverage of PAM currently so I think adding that here might be out of scope for this PR though it would be very beneficial.

**Comment by AndrewJackson2020** [comment]
@JelteF Fixed the macOS tests by scaling back the run_with_config usage as described in #1138.

Let me know if this solution works for you or if it needs to be reworked more.

**Comment by AndrewJackson2020** [comment]
@JelteF Anything else I can do here to help this PR get merged?

**Comment by JelteF** [comment]
No, it's blocked on me finding some time to review PgBouncer PRs again. Busy busy.

File: src/client.c
Comment (JelteF) [inline src/client.c:432]: That comment seems out of date now.
```suggestion
```

File: src/client.c
Comment (JelteF) [inline src/client.c:435]: I agree this is not pretty. But given the amount of counting bugs found in earlier versions, I prefer this way because this one is obviously correct and will stay correct in the future. I added a check for user_connection_counted in the decrement path too.

**Comment by AndrewJackson2020** [comment]
@JelteF anything else you want done with this PR before merging?

**Comment by JelteF** [comment]
I don't remember why I didn't press merge yet, maybe because CI was failing for unrelated reasons.

**Pull request merged**
